src/contracts/core/StakeRootCompendium.sol

```diff
@@ -234,226 +202,155 @@ contract StakeRootCompendium is StakeRootCompendiumStorage {
+            "StakeRootCompendium._addToStakeTree: operator set does not meet minimum deposit balance"
         );
+        operatorSets.push(operatorSet);
+        // note: operatorSetIndexPlusOne is 1 indexed
+        operatorSetToOneBasedIndex[operatorSet.avs][operatorSet.operatorSetId] = uint208(operatorSets.length);
     }
 
-    function _getPreviousCumulativeCharges(uint32 timestamp) internal view returns(uint224, uint224) {
-        return (
-            cumulativeContantChargeSnapshot.upperLookup(timestamp),
-            cumulativeLinearChargeSnapshot.upperLookup(timestamp)
+    function _removeOperatorSetFromStakeTree(IAVSDirectory.OperatorSet memory operatorSet) internal {
+        uint256 operatorSetOneBasedIndex = operatorSetToOneBasedIndex[operatorSet.avs][operatorSet.operatorSetId];
+        require(
+            operatorSetOneBasedIndex != 0, 
+            "StakeRootCompendium._addToStakeTree: operator set already removed from stake tree"
         );
-    }
 
-    function _getCurrentCumulativeCharge(Checkpoints.History storage cumulativeChargeSnapshot, uint64 currentCharge, uint32 latestCalculationTimestamp) internal view returns (uint224) {
-        (bool exists, uint32 latestSnapshotTimestamp, uint224 latestSnapshottedCumulativeCharge) = cumulativeChargeSnapshot.latestCheckpoint();
-        // return the latest snapshot if it has been accounted for
-        if(latestSnapshotTimestamp == latestCalculationTimestamp) {
-            return latestSnapshottedCumulativeCharge;
-        }
-
-        // use "now" if this is the first time the charge is set
-        if (!exists) {
-            latestSnapshotTimestamp = latestCalculationTimestamp;
-            latestSnapshottedCumulativeCharge = 0;
-        }
+        IAVSDirectory.OperatorSet memory substituteOperatorSet = operatorSets[operatorSets.length - 1];
+        // note: operatorSetIndex is 1 indexed so operatorSetOneBasedIndex - 1 is strictly non-negative
+        operatorSets[operatorSetOneBasedIndex - 1] = substituteOperatorSet;
+        operatorSets.pop();
 
-        // keep track of the cumulative charge that would have been charged since the last change
-        // this is: 
-        // latestSnapshottedCumulativeCharge + proofs since latest snapshot * (charge / proof)
-        // latestSnapshottedCumulativeCharge + (latestCalculationTimestamp - latestSnapshotTimestamp) * (1 proof / proofInterval time) * (charge / proof)
-        // latestSnapshottedCumulativeCharge + (charge / proof) * (time since snapshot) * (1 proof / proofInterval time)
-        return latestSnapshottedCumulativeCharge + uint224(currentCharge * (latestCalculationTimestamp - latestSnapshotTimestamp) / proofInterval);
+        // update the index of the operator sets
+        operatorSetToOneBasedIndex[msg.sender][operatorSet.operatorSetId] = 0;
+        operatorSetToOneBasedIndex[msg.sender][substituteOperatorSet.operatorSetId] = operatorSetOneBasedIndex;
```

Comment:
this should not be msg.sender, it should be the `operatorSet.avs` and `substituteOperatorSet.avs` respectively

Reply:
yeah, definitely